### ISSUE TITLE
quality(epic): make lint and type gates honest, then reduce comment and monolith debt

### ISSUE THREAD

**Issue by pablomoli** [pablomoli/ytk#122]
## Severity

**High engineering risk, not a current production incident.**

The repository now has useful linting, formatting, and type-checking tools, but the default green signals cover much less code than their names imply. This matters because the highest-risk files are also the least structurally constrained: `web/src/lib/mapRenderer.ts`, `web/src/routes/settings.tsx`, `ytk/cli.py`, and `ytk/ui/hub.py`.

This issue is the repository-wide follow-up to #101 / PR #110. Those changes were a valid first ratchet, but #101 was closed under the language of "strict typing/linting" even though PR #110 explicitly deferred the largest chunks. The comment-cleanup acceptance item in #101 also does not appear in PR #110's delivered scope.

## Why #101 was implemented narrowly

PR #110 was explicitly constrained to **no behavior changes** and optimized for a fast local pre-commit hook:

- Ruff began with 395 violations across 166 Python files.
- 233 were auto-fixed and six defects were fixed manually.
- 170 findings were globally ignored because they were considered codebase-specific or behavior-changing.
- Basic Pyright across the package produced 308 errors at the time.
- Strict Pyright on `ytk/ridges.py` produced 328 diagnostics, 325 from NumPy `Unknown` types.
- `noUncheckedIndexedAccess` produced 289 frontend errors, including 91 in `mapRenderer.ts`.

The implementation therefore chose a small gate that was already green:

- Ruff and Ruff format on staged Python files under `ytk/`, `scripts/`, and `tests/`.
- Pyright in basic mode on `ytk/ridges.py` only, triggered only when that file changes.
- Frontend lint/type commands run manually, not in the pre-commit hook.

That was a reasonable first ratchet. The problem is that the deferred work was never turned into a follow-up gate-expansion issue, while #101 was closed as if the strict gate were complete.

### Python lint and format

Configured in `pyproject.toml`:

```toml
[tool.ruff.lint]
select = ["E", "F", "I", "UP", "B", "SIM", "RUF"]
```

The pre-commit hook runs Ruff check and Ruff format only on staged `.py` files matching:

```text
^(ytk|scripts|tests)/.*\.py$
```

What this catches well:

- syntax and undefined-name failures
- unused imports/variables
- import sorting
- modernization rules
- a useful subset of bugbear and simplification rules
- formatting drift

What it does not currently cover:

- `experiments/` and `labs/` in the hook
- whole-repository regressions caused by config or shared imports
- complexity (`C901`, `PLR0912`, `PLR0915`)
- security-oriented checks
- broad exception policy
- architecture/module-size limits

The global ignore list also suppresses rules everywhere, even when the stated justification applies only to a subset of files. Example: `E402` is described as required by bootstrap scripts, but isolated Ruff finds production violations in:

- `ytk/cli.py`
- `ytk/store.py`
- `ytk/ui/server.py`

Current full Ruff check including experiments finds seven failures, while the staged hook cannot see them.

### Python types

Current configuration:

```toml
[tool.pyright]
include = ["ytk/ridges.py"]
typeCheckingMode = "basic"
```

The hook runs Pyright only when `ytk/ridges.py` is staged. Therefore:

- `uv run pyright` reports `0 errors`
- `uv run pyright ytk` currently reports **120 errors across 50 files**
- 60 are in `ytk/store.py`
- the largest categories are argument-type and unsafe-optional errors

The configured zero is real for `ridges.py`, but it is not a backend-wide type gate.

### Frontend

The frontend has:

- TypeScript strict mode plus `exactOptionalPropertyTypes`, `noImplicitOverride`, and `noImplicitReturns`
- type-aware `vp lint`
- `vp test`
- `vp build`

However:

- none run from the repository pre-commit hook
- no tracked GitHub Actions workflow runs them remotely
- `vp lint` still exits zero with two `react-hooks/exhaustive-deps` warnings
- route test files are repeatedly inspected as TanStack routes
- `noUncheckedIndexedAccess` remains disabled after producing 289 errors

### CI

There are no tracked `.github/workflows` files. The only automatic enforcement is a clone-local hook configured through `core.hooksPath`, and it can be bypassed with `git commit --no-verify`.

The retrieval evaluation hook is valuable and should remain separate; it protects search quality, not general code quality.

## Audit findings

The repository audit covered 345 hand-written code/config files.

### Structural hotspots

- `web/src/lib/mapRenderer.ts`
  - `mountMapRenderer`: 1,651 lines, 204 branch points
- `web/src/routes/settings.tsx`
  - `SettingsPage`: 553 lines
- `ytk/cli.py`
  - 3,006-line module
  - `triage`: 187 lines / 30 branch points
  - several other commands exceed 100 lines
- `ytk/ui/hub.py`
  - `refresh_sources`: 169 lines / 28 branch points

An isolated Ruff complexity report finds 118 complexity/size violations across production code, scripts, experiments, and tests.

### Error handling

There are 132 `except Exception` blocks across Python source and scripts:

- 28 in `ytk/ui/hub.py`
- 16 in `ytk/cli.py`
- 8 in `ytk/ui/server.py`

Several API endpoints return `detail=str(exc)`, coupling user-visible responses to arbitrary internal exception text.

### Comment debt

The problem is concentrated, not repository-wide. The following comments should be removed, shortened, or moved to decision/debugging documents:

- `pyproject.toml`: lint-adoption history, previous counts, and pass constraints embedded in configuration
- `scripts/build_map.py`: 17-line corpus-specific history above `THEME_FLOOR`
- `web/src/lib/mapRenderer.ts`: failed bloom hypothesis and unexplained 1.8x discrepancy embedded in a shader string
- `web/vite.config.ts`: exact warning counts from a past lint run
- stale `PROTOTYPE`, workshop, experiment-number, and "throwaway" headers in grove modules
- duplicated native-dialog/StrictMode explanation in `ConfirmDialog`, `NoteViewer`, and both test files
- conversational failure-history comments such as "it took the hub down every time we tried it"

Do not indiscriminately remove comments. Keep short comments that preserve:

- WebGL limits and precision constraints
- MPS/encoder memory invariants
- embedding epoch and truncation behavior
- mathematical algorithm invariants
- benchmark methodology that prevents invalid measurements

### 1. Make the gates honest before refactoring

- Add a tracked CI workflow for Python lint/format, backend type checking, frontend lint/tests/build, and fast Python tests.
- Keep the local hook fast, but make CI authoritative and repository-wide.
- Include `experiments/` and `labs/`, or explicitly classify and exclude them with documented per-path policy.
- Replace global Ruff ignores with narrow `per-file-ignores` where the reason is path-specific.
- Fix the current seven Ruff failures.
- Fix the two frontend hook warnings and configure TanStack route-test exclusions.

### 2. Widen types incrementally

- First make `uv run pyright ytk` pass in `basic` mode.
- Fix real optional/argument mismatches before adding suppressions.
- Ratchet Pyright module by module from basic to strict.
- Enable `noUncheckedIndexedAccess` by subsystem, starting with small libraries and tests before `mapRenderer.ts`.
- Publish the exact checked scope in command names or CI job names so "green" cannot be mistaken for repository-wide coverage.

### 3. Perform a comment-only cleanup

- Delete stale labels and conversational/session-history prose.
- Move measurements and decision history into existing docs/decision/debugging artifacts.
- Leave one- to three-line invariant comments beside code.
- Add a short comment policy to `AGENTS.md`: explain why/invariants, not development history; no issue-phase narration; no duplicated explanations.

Keep this change behavior-neutral and review it separately from refactors.

### 4. Consolidate duplicated lifecycle logic

- Extract the native `<dialog>` lifecycle into one hook or primitive.
- Share the jsdom dialog polyfill/test helper.
- Remove the four copies of the StrictMode explanation.

### 5. Split monoliths under the expanded gates

Suggested risk order:

1. `refresh_sources`: provider adapters plus one orchestrator
2. `SettingsPage`: one component per settings section
3. `ytk/cli.py`: commands become thin adapters over service functions
4. `mountMapRenderer`: extract shader/program resources, terrain, picking, bloom, animation state, labels, and input handling

`mountMapRenderer` goes last because it is the highest-risk refactor and depends on browser/render witnesses that static tools cannot provide.

### 6. Standardize error boundaries

- Catch expected exception types inside services.
- Keep one broad catch only at CLI/job/API boundaries.
- Log internal details there and return stable user-facing errors.
- Remove raw `detail=str(exc)` responses where exceptions may contain paths, URLs, or credentials.

## Acceptance criteria

- [ ] Tracked CI exists and runs on every PR.
- [ ] `uv run ruff check ytk scripts tests experiments labs` passes.
- [ ] `uv run ruff format --check ytk scripts tests experiments labs` passes.
- [ ] `uv run pyright ytk` passes in basic mode before strictness is widened.
- [ ] Frontend lint exits with zero errors and zero warnings.
- [ ] Frontend tests and production build pass in CI.
- [ ] Global Ruff ignores that apply only to scripts become per-file ignores.
- [ ] Deferred strictness work is represented by checked scopes and explicit follow-up tasks.
- [ ] Historical/conversational comments listed above are removed or relocated.
- [ ] `ConfirmDialog` and `NoteViewer` share dialog lifecycle infrastructure.
- [ ] Each monolith has a decomposition plan and characterization tests before extraction.

Refs: #101, #107, PR #110.

**Comment by pablomoli** [comment]
## Phase 2, bullet 1 done: `uv run pyright ytk` is at 0 errors

Commits: e0864ad, 2093268, 2c9639c, ec21f52. 120 -> 0. 796 tests pass throughout.

**60 of 120 were one root cause.** `store.py` is the only module touching Chroma directly, and Chroma types every Get/Query field `List[X] | None` (`include=` decides what returns) and every metadata value as a str|int|float|bool|SparseVector|None union. ytk always requests what it reads and only writes strings — real knowledge, living in the developer's head. Three accessors (`chroma_field`/`meta_str`/`meta_float`) state it once and are shared with graph, tags, retrieval_gate and the hub server.

**No `Any` anywhere in the package**, per review. Where an escape hatch was genuinely needed, it is a Protocol + one `cast` at the load boundary. Annotating the transformers objects `PreTrainedModel`/`ProcessorMixin` was tried first and measured *worse* — 4 errors became 19, and clearing them needed a `type: ignore` per call site, which hides real errors too. The Protocols name the exact surface used, so everything downstream stays checked. `workboard.py`'s `_load_json` went from `-> Any` to a recursive `JsonValue` with narrowing helpers; verified against live data (66 items, 3 in progress, identical output).

Four narrow rule-named `type: ignore` lines remain, each where a third-party stub contradicts its own runtime (yt_dlp `_Params`, sklearn `n_init`, googleapiclient's dynamic `Resource`). One line each; no untyped value flows onward.

### Real defects found, not just annotation gaps
- `reels.py` — an empty-string `thread_id` fell through both `if thread_id:` and `if thread_id is None:`, leaving `messages` unset and the loop iterating `None`.
- `synthesis.py` — `raise grounding_error` raised `None` when no attempt had failed.
- `reddit_feed.py` — `_get` was annotated `-> dict`, but a post permalink returns a **list**.
- `books_match.py` — unchecked `re.search(...).group(1)` on container.xml.
- `sdk.py` — `add_dirs: list[str | Path]` was unsatisfiable from a `list[Path]` caller (list invariance).

### The gate, which is the actual point
`[tool.pyright]` included exactly one file, so `uv run pyright` reported a truthful-looking `0 errors` for a scope nobody assumes — the same failure mode as #114. Now `include = ["ytk"]`, and the pre-commit hook runs pyright on any staged `ytk/*.py` rather than only `ridges.py`. Fixing the errors without this just resets a counter that drifts back.

Still `basic`. Strict on ridges.py alone gives 328 diagnostics, all but 3 the `reportUnknown*` family from numpy's `ndarray[Unknown, Unknown]` — a separate per-module ratchet.

### Frontend note
There is **no hand-written `any` in web/src**. All 12 hits are in `routeTree.gen.ts`, which is TanStack codegen carrying `@ts-nocheck` and regenerated on every route change.

### Measured, for the rest of this epic
Stricter ruff on `ytk/` (not adopted — needs a scope decision):

| family | hits | | family | hits |
|---|---|---|---|---|
| PL (pylint) | 470 | | S (security) | 97 |
| D (docstrings) | 297 | | BLE (blind except) | 98 |
| ANN (annotations) | 211 | | TRY | 93 |
| EM | 84 | | SLF | 34 |
| DTZ | 33 | | N | 26 |
| C90 (complexity) | 22 | | PTH | 11 |

Currently-deferred rules still real: B904 44, B905 29, SIM105 5, B023 4, RUF005 2.

`BLE` (98) and `C90` (22) map directly to this issue's own error-handling and monolith findings; `S` (97) is unreviewed security signal. `D`/`ANN` (508 combined) are style mandates worth deciding on deliberately rather than absorbing.

Remaining phases untouched: CI workflow, ruff widening + per-file-ignores, comment cleanup, dialog-lifecycle dedup.

**Comment by pablomoli** [comment]
## Ruff families adopted (59aefb4)

Phase 1's lint half, alongside the type work already reported. **Selected and every finding fixed** — nothing is selected-but-ignored:

`C4`, `PIE`, `RET`, `PERF`, `FURB`, `LOG`, `G`, `S324`. 67 findings, 52 autofixed, the rest by hand. `LOG` and `G` were already clean and are now held that way.

`S324` was the security rule worth taking whole: six truncated SHA-1 digests used as cache keys and note filenames, plus one in a test helper mirroring the hub's key. `usedforsecurity=False` states the intent and leaves the digest bytes identical.

### Not adopted — decisions, with the numbers

Measured across `ytk/` + `scripts/` + `tests/` first, because the headline counts are misleading:

| family | total | what it actually is |
|---|---|---|
| `ANN` | 2569 | Duplicates the pyright strict ratchet (c189be1). `reportMissingParameterType` already covers it, per-module with an opt-out list. A second mechanism at a different scope fights it rather than tightening it. |
| `S` | 1924 | 1812 are `S101` (assert) in `tests/`. `ytk/` has 97, and `S603`/`S607`/`S310` are ytk shelling out and fetching arbitrary URLs by design — 68 noqa comments for no safety gain. |
| `PL` | 1160 | 321 are `PLC0415`, and ytk defers heavy imports deliberately: `graph.py` imports `store` inside `build_graph` so drawing a graph does not load the encoder stack. |
| `D` | 1409 | Docstring house style — wants a deliberate yes/no, not a default. |

`PERF401` is ignored with its reason: the 16 loops it targets carry guard clauses, which is why ruff will not autofix them; hand-rolling the comprehension trades a readable loop for a dense one.

All counts and reasons now live in `[tool.ruff.lint]` so "not adopted" cannot be misread as "not considered".

### Structural, per this issue's own asks

- **`per-file-ignores` replaces reaching for a global ignore.** `tests/` and `scripts/` waive `S101`/`S108`; production code does not. This was an explicit ask in the issue.
- **The pre-commit hook had a hole:** it linted only *staged* files, so editing `pyproject.toml` — which changes what every file is judged by — could pass while breaking the repo. It now lints `ytk/ scripts/ tests/` whenever the config moves.

ruff clean, pyright 0 errors, 796 tests pass.

---

## Status of this epic

**Done:** phase 2 bullet 1 (backend types at 0, strict ratchet, scope widened) and the adoptable half of phase 1's lint widening.

**Still open, unchanged:**
1. CI workflow — there are still no tracked `.github/workflows`; the only enforcement is a clone-local hook bypassable with `--no-verify`.
2. `BLE` 119 + `TRY` 145 + `S110` 17 — the broad-exception work. Related: this sweep found `refresh_sources` swallowing per-source exceptions is what hid #114 for months.
3. `C901` 22 + `PLR09xx` 69 — the monolith work (`mapRenderer.ts`, `cli.py`, `hub.py`).
4. Comment-only cleanup.
5. Dialog-lifecycle dedup.
6. Frontend: `vp lint` warnings, `noUncheckedIndexedAccess`.

Both remaining lint families are code changes rather than config, and each wants its own reviewed pass.

**Comment by pablomoli** [comment]
Implemented and verified on `issue-122-honest-gates` through `b922757`.

Acceptance ledger:

- CI/PR criterion intentionally superseded by the owner decision in this ticket: this single-contributor repository does not use PRs, so `just check` is the authoritative full local gate. No GitHub Action or PR was added.
- `uv run --extra dev ruff check ytk scripts tests experiments labs`: pass.
- `uv run --extra dev ruff format --check ytk scripts tests experiments labs`: pass; 212 files formatted.
- Pyright now checks all of `ytk` in strict mode: 0 errors, 0 warnings.
- Frontend lint exits cleanly with no warnings.
- Chromium frontend tests: 59 files, 278 tests passed.
- Production frontend build: pass. The existing 1.35 MB chunk warning remains visible as separate build-performance debt.
- Ruff exceptions whose justification is path-specific are path-scoped; E402 is limited to exact initialization-sensitive modules, scripts, and one legacy test module.
- Deferred strictness and extraction scope is explicit in the checked commands, `CLAUDE.md`, and architecture documents.
- Historical and conversational comments were removed or rewritten as current invariants; `AGENTS.md` now states the policy.
- `ConfirmDialog`, `NoteViewer`, and `QueueItemViewer` share `useModalDialog` lifecycle infrastructure.
- `refresh_sources` now delegates to characterized typed provider adapters; settings rendering is separated from the draft/save transaction with characterization coverage.
- Witness-first decomposition plans cover `ytk/cli.py` and `web/src/lib/mapRenderer.ts` before extraction.
- The retired simulated DOM package is absent as an installed or resolved dependency. Only Vitest upstream optional-peer metadata remains in the lockfile; `autoInstallPeers: false` plus repository hygiene coverage prevents resolution or authored references.
- Complete `just check`: 859 Python tests passed, 1 deselected; 278 frontend tests passed; build passed.
- `git diff --check`: clean.

Closeout: `docs/session-045-brief.md`; vault mirror, hot note, and index updated. No PR was created.

**Comment by pablomoli** [comment]
Landed on `master` in merge commit `9f1e20d` after integrating the three newer master commits and rebuilding the combined tracked bundle. Post-merge `just check` passed with 865 Python tests and 296 Chromium tests.

Residual work is parked and ordered on Project 3:

- #108 — stop committing `web/dist` (P1, Ready, Order 2; next)
- #138 — persist inbox profile reroll position (P1, Ready, Order 10.7)
- #142 — route-level JavaScript bundle splitting (P2, Ready, Order 19)
- #141 — CLI service extraction (P2, Ready, Order 24)
- #143 — renderer resource-owner decomposition (P2, Ready, Order 61)
